File: android/src/main/java/com/axity/plugin/certificate/CertificateInfo.java
package com.axity.plugin.certificate;

import android.content.Context;
import android.content.res.AssetManager;
import android.util.Log;

import org.json.JSONArray;

import java.io.Console;
import java.io.IOException;
import java.io.InputStream;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.CertificateExpiredException;
import java.security.cert.CertificateFactory;
import java.security.cert.CertificateNotYetValidException;
import java.security.cert.X509Certificate;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.TimeZone;

public class CertificateInfo {
  private final Context context;

  private final CertificateInfoConfig config;

  public CertificateInfo(Context context, CertificateInfoConfig config) {
    this.context = context;
    this.config = config;
  }

  public Date getExpirationDate(String filename) throws Exception {
    if (filename == null) {
      throw new CertificateInfoException(CertificateInfoError.FILE_NAME_EMPTY);
    }

    X509Certificate cert = openCertificate(filename);

    return cert.getNotAfter();
  }

  public List<Date> getExpirationDates() throws Exception {
    List<Date> expirationDates = new ArrayList<>();

    List<X509Certificate> certificates = openCertificates();

    for (X509Certificate certificate : certificates) {
      expirationDates.add(certificate.getNotAfter());
    }

    return expirationDates;
  }

  public boolean checkValidation(String date) throws Exception {

    Date dateToCompare = getToday();

    if(!date.isEmpty()){
      dateToCompare = parseDate(date);
    }

    List<X509Certificate> certificates = openCertificates();
    try{

      for (X509Certificate certificate : certificates) {
          certificate.checkValidity(dateToCompare);
      }

      return true;

    }catch(CertificateExpiredException | CertificateNotYetValidException e){
      Log.i("CERTIFICATE_INFO", "Some certificate is expired" + e.getMessage());
      return false;
    }
  }

  public String formatDateToISO(Date dt) {
    SimpleDateFormat isoFormat = new SimpleDateFormat(config.getDateFormat(), java.util.Locale.getDefault());
    return isoFormat.format(dt);
  }

  private Date parseDate(String date){
    SimpleDateFormat sdf = new SimpleDateFormat(config.PARSE_DATE_FORMAT, Locale.getDefault());
    sdf.setTimeZone(TimeZone.getTimeZone(config.TIME_ZONE));

    try {
      return sdf.parse(date);
    } catch (ParseException e) {
      Log.i("CERTIFICATE_INFO", "WARNING: Error parsing date, returning current date");
      return getToday();
    }

  }

  private List<X509Certificate> openCertificates() throws Exception {
    List<X509Certificate> certificates = new ArrayList<>();
    AssetManager assetManager = context.getAssets();
    String[] certificateFiles = assetManager.list(config.getCertificatePath());

    if (certificateFiles == null) {
      throw new CertificateInfoException(CertificateInfoError.READ_DIRECTORY_FILES);
    }

    for (String fileName : certificateFiles) {
      if (fileName.endsWith(config.CERTIFICATE_EXTENSION)) {
        X509Certificate certificate = openCertificate(fileName);
        certificates.add(certificate);
      }
    }

    return certificates;
  }

  private X509Certificate openCertificate(String filename) throws Exception {
    try {
      AssetManager assetManager = context.getAssets();
      InputStream inputStream = assetManager.open(config.getCertificatePath() + "/" + filename);

      CertificateFactory cf = CertificateFactory.getInstance(config.CERTIFICATE_FORMAT);
      Certificate cert = cf.generateCertificate(inputStream);
      inputStream.close();

      if (cert instanceof X509Certificate x509Cert) {
        return x509Cert;
      } else {
        throw new CertificateInfoException(CertificateInfoError.CERTIFICATE_FORMAT_SUPPORT);
      }

    } catch (IOException e) {
      throw new CertificateInfoException(CertificateInfoError.FILE_NOT_FOUND, filename);
    }catch (CertificateException e){
      throw new CertificateInfoException(CertificateInfoError.READ_FILE_ERROR, filename);
    }

  }

  private Date getToday(){
    Calendar today = Calendar.getInstance();
    today.set(Calendar.HOUR_OF_DAY, 0);
    today.set(Calendar.MINUTE, 0);
    today.set(Calendar.SECOND, 0);
    today.set(Calendar.MILLISECOND, 0);

    return today.getTime();
  }

}

File: ios/Sources/CertificateInfoPlugin/CertificateInfo.swift
import Foundation
import Security
import ASN1Decoder




@objc public class CertificateInfo: NSObject {
  
  var config: CertificateInfoConfig = CertificateInfoConfig()
  
  public init(config: CertificateInfoConfig) {
    self.config = config;
  }
  
  @objc public func getExpirationDate(from filename: String) throws -> Date {
    
    if(filename.isEmpty){
      throw CertificateInforError.fileNameEmpty
    }
    
    let cerdata = try openCertificate(filename:filename)
    
    
    if let expirationDate = cerdata.notAfter {
      return expirationDate

    } else {
      throw CertificateInforError.extractDateError
    }
  }
  
  @objc public func getExpirationDates() throws -> [Date] {
    var expirationDates: [Date] = []
    
    let cerFiles = try openCertificates()
    
    for cerFile in cerFiles {
      if let expirationDate = cerFile.notAfter {
        expirationDates.append(expirationDate)
      }
    }
    
    return expirationDates
  }
  
  @objc public func checkValidation(date: String?) throws -> NSNumber {
    let today = Calendar.current.startOfDay(for: Date());
    
    var dayToCompare: Date
    if let dateString = date {
      dayToCompare = parseDate(date: dateString)
    }else{
      dayToCompare = today
    }
    
    
    let certificates = try openCertificates()
    
    for certificate in certificates {
      
      if(!certificate.checkValidity(dayToCompare)){
          if let exp = certificate.notAfter {
              print("Some certificate is expired " + formatDateToISO(exp) )
          }
        return NSNumber(value: false)
      }
    }
    
    return NSNumber(value: true)
  }

  
  public func formatDateToISO(_ date: Date) -> String {

    let isoFormatter = ISO8601DateFormatter()
    let isoDateString = isoFormatter.string(from: date)

    return isoDateString
  }
  
  private func parseDate(date: String) -> Date {

    let isoDateFormatter = ISO8601DateFormatter()
    isoDateFormatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let date = isoDateFormatter.date(from: date) {
       return date
    }
    
    print("WARNING: Error parsing date, returning current date")
    return Calendar.current.startOfDay(for: Date())
  }
  
  
  private func openCertificates() throws -> [X509Certificate] {
    var certificates: [X509Certificate] = []
    let fileManager = FileManager.default
    let bundlePath = "\(Bundle.main.resourcePath!)/\(self.config.certificatePath)"

    guard let directoryContents = try? fileManager.contentsOfDirectory(atPath: bundlePath) else {
      throw CertificateInforError.readDirectoryFiles
    }
    
    let cerFiles = directoryContents.filter { $0.hasSuffix(self.config.CERTIFICATE_EXTENSION) }
  
    for file in cerFiles {
      let certificate = try self.openCertificate(filename: file)
      certificates.append( certificate )
    }
    
    return certificates
  }
  
  private func openCertificate(filename: String) throws -> X509Certificate {
    
    let fileWithoutExtension = (filename as NSString).deletingPathExtension
    let cerPath = "\(self.config.certificatePath)/\(fileWithoutExtension)"
    
    guard let certificatePath = Bundle.main.path(forResource:cerPath, ofType: "cer") else {
      throw CertificateInforError.fileNotFound(filename)
    }
    
    guard let certificateData = try? Data(contentsOf: URL(fileURLWithPath: certificatePath)) else {
      throw CertificateInforError.readFileError(filename)
    }
    
    guard let cerX509 = try? X509Certificate(data: certificateData) else {
      throw CertificateInforError.certificateFormatSupport
    }
    
    return cerX509
  }
}
